File: backend/src/routes/files.ts
import { Router, Request, Response, NextFunction } from 'express';
import archiver from 'archiver';
import * as ShareService from '../services/ShareService';
import * as StorageService from '../services/StorageService';
import { normalizeUID, isValidUID } from '../services/UIDService';
import { validateSessionToken } from '../services/PasswordService';
import { retrievalLimiter } from '../middleware/rateLimiter';
import { createError } from '../middleware/errorHandler';

const router = Router();

// ─── GET /api/files/:id/download — Presigned download proxy ───
router.get(
  '/:id/download',
  retrievalLimiter,
  async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
      const file = await ShareService.getFileForDownload(req.params.id);
      if (!file) {
        throw createError('File not found or share has expired', 404);
      }

      // If the share is private, check session token
      if (file.isPrivate) {
        const token = req.headers['x-session-token'] as string;
        if (!token || !validateSessionToken(token, file.shareUid)) {
          throw createError('Authentication required', 401);
        }
      }

      // Generate presigned URL (60s TTL) and redirect
      const url = await StorageService.getPresignedDownloadUrl(file.storageKey, file.filename);
      res.redirect(url);
    } catch (error) {
      next(error);
    }
  }
);

export default router;

File: backend/app/routers/files.py
"""
files.py — File download route via presigned URL redirect.
Replaces: backend/src/routes/files.ts
"""

from fastapi import APIRouter, Depends, Request, Header
from fastapi.responses import JSONResponse, RedirectResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.services import share_service, storage_service, password_service
from app.middleware.rate_limiter import limiter, RETRIEVAL_LIMIT

router = APIRouter(prefix="/api/files")


@router.get("/{file_id}/download")
@limiter.limit(RETRIEVAL_LIMIT)
async def download_file(
    request: Request,
    file_id: str,
    db: AsyncSession = Depends(get_db),
    x_session_token: str | None = Header(default=None, alias="X-Session-Token"),
):
    file = await share_service.get_file_for_download(db, file_id)
    if not file:
        return JSONResponse(
            status_code=404,
            content={"success": False, "error": "File not found or share has expired"},
        )

    # If the share is private, check session token
    if file["is_private"]:
        token = x_session_token or request.query_params.get("token")
        if not token or not password_service.validate_session_token(token, file["share_uid"]):
            return JSONResponse(
                status_code=401,
                content={"success": False, "error": "Authentication required"},
            )

    # Generate presigned URL (60s TTL) and redirect
    url = await storage_service.get_presigned_download_url(
        file["storage_key"], file["filename"]
    )
    return RedirectResponse(url=url)
